checkstyle/src/main.rs

```diff
@@ -28,78 +25,76 @@ fn read_file_content(file_path: &Path) -> String {
     fs::read_to_string(file_path).expect("Failed to read file")
 }
 
-fn check_matching(file: &str, regex: &str) -> Option<bool> {
+fn check_direct_matching(file_lines: Vec<&str>, regex_lines: Vec<&str>) -> Option<bool> {
+    if file_lines.len() < regex_lines.len() {
+        return Some(false);
+    }
+    let together = file_lines.iter().zip(regex_lines.iter());
+    for (file_line, regex_line) in together {
+        let re = Regex::new(regex_line).ok()?;
+        if !re.is_match(file_line) {
+            return Some(false);
+        }
+    }
+    Some(true)
+}
+
+fn check_matching(file: &str, regex: &str) -> bool {
     let file_lines: Vec<&str> = file.lines().collect();
     let regex_lines: Vec<&str> = regex.lines().collect();
-    let mut file_id = 0;
-    let mut regex_id = 0;
-    while file_id < file_lines.len() && regex_id < regex_lines.len() {
-        let regex_line = regex_lines[regex_id];
-        let file_line = file_lines[file_id];
-        let re = Regex::new(regex_line).ok()?;
-        if re.is_match(file_line) {
-            file_id += 1;
-            regex_id += 1;
-        } else {
-            file_id += 1;
+    for i in 0..file_lines.len() {
+        let suffix = &file_lines[i..];
+        if check_direct_matching(suffix.to_vec(), regex_lines.to_vec()).unwrap_or(false) {
+            return true;
         }
     }
-    Some(regex_id == regex_lines.len())
+    return false;
 }
 
-fn list_files_in_folder(folder_path: &Path) -> Result<Vec<String>, std::io::Error> {
+fn list_files_in_folder(folder_path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
     let mut file_paths = Vec::new();
 
     for entry in fs::read_dir(folder_path)? {
         let entry = entry?;
         let path = entry.path();
 
         if path.is_file() {
-            if let Some(file_name) = path.file_name() {
-                if let Some(file_name_str) = file_name.to_str() {
-                    file_paths.push(file_name_str.to_string());
-                }
+            file_paths.push(path.clone());
+        } else if path.is_dir() {
+            if let Ok(sub_paths) = list_files_in_folder(&path) {
+                file_paths.extend(sub_paths);
```

Comment:
Let's make sure to check whether our `dir` is a symlink to avoid creating infinite loops.